samples/boards/stm32/pm_lptim/src/main.c

```diff
@@ -0,0 +1,119 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+
+#include <string.h>
+#include <soc.h>
+#include <device.h>
+#include <drivers/gpio.h>
+#include <sys/printk.h>
+
+struct device *gpio_bn, *gpio_led;
+
+bool ready_to_sleep;
+
+/* in msec */
+/* setting to K_TICKS_FOREVER willl activate the deepsleep mode */
+#define SLEEP_TIME 3000
+
+#define DEMO_DESCRIPTION	\
+	"Demo Description\n"	\
+	"Application is demonstrating the low power" \
+	"with the LPTIM used as systick.\n" \
+	"The system goes to sleep mode on k_sleep() when pressing the button\n" \
+	"(led off) and wakes up by the lptim timeout expiration (led on)\n\n" \
+
+/* The devicetree node identifier for the "sw0" alias. */
+#define SW0_NODE DT_ALIAS(sw0)
+/* GPIO for the User push button */
+#if DT_NODE_HAS_STATUS(SW0_NODE, okay)
+#define SW	DT_GPIO_LABEL(SW0_NODE, gpios)
+#define PIN	DT_GPIO_PIN(SW0_NODE, gpios)
+#if DT_PHA_HAS_CELL(SW0_NODE, gpios, flags)
+#define SW0_FLAGS DT_GPIO_FLAGS(SW0_NODE, gpios)
+#else
+#define SW0_FLAGS 0
+#endif
+#else
+/* A build error here means your board isn't set up to input a button. */
+#error "Unsupported board: sw0 devicetree alias is not defined"
+#endif
+
+/* The devicetree node identifier for the "led0" alias. */
+#define LED0_NODE DT_ALIAS(led0)
+/* GPIO for the Led */
+#if DT_NODE_HAS_STATUS(LED0_NODE, okay)
+#define LED0	DT_GPIO_LABEL(LED0_NODE, gpios)
+#define PIN0	DT_GPIO_PIN(LED0_NODE, gpios)
+#if DT_PHA_HAS_CELL(LED0_NODE, gpios, flags)
+#define LED0_FLAGS DT_GPIO_FLAGS(LED0_NODE, gpios)
+#else
+#define LED0_FLAGS 0
+#endif
+#else
+/* A build error here means your board isn't set up to blink an LED. */
+#error "Unsupported board: led0 devicetree alias is not defined"
+#endif
+
+static struct gpio_callback gpio_cb;
+
+void button_pressed(struct device *dev, struct gpio_callback *cb,
+		    uint32_t pins)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(pins);
+	ARG_UNUSED(cb);
+
+	/* Set pin to HIGH/LOW */
+	gpio_pin_set(gpio_led, PIN0, 0);
+	ready_to_sleep = true;
+}
+
+/* Application main Thread */
+void main(void)
+{
+
+	printk("\n\n*** Low power Management Demo on %s ***\n", CONFIG_BOARD);
+	printk(DEMO_DESCRIPTION);
+
+	ready_to_sleep = false;
+	gpio_bn = device_get_binding(SW);
+
+	/* Configure Button 1 to wakeup from sleep mode */
+	gpio_pin_configure(gpio_bn, PIN, SW0_FLAGS | GPIO_INPUT);
+	gpio_pin_interrupt_configure(gpio_bn, PIN, GPIO_INT_EDGE_TO_ACTIVE);
+	gpio_init_callback(&gpio_cb, button_pressed, BIT(PIN));
+	gpio_add_callback(gpio_bn, &gpio_cb);
+
+	/* Configure LEDs */
+	gpio_led = device_get_binding(LED0);
+	if (gpio_led == NULL) {
+		return;
+	}
+
+	if (gpio_pin_configure(gpio_led, PIN0,
+			GPIO_OUTPUT_ACTIVE | LED0_FLAGS) < 0) {
+		return;
+	}
+
+	/*
+	 * Start the demo.
+	 */
+
+	while (1) {
+		if (ready_to_sleep == true) {
+			gpio_pin_set(gpio_led, PIN0, 0);
+			ready_to_sleep = false;
+			k_msleep(SLEEP_TIME);
+		}
+
+		gpio_pin_set(gpio_led, PIN0, 1);
+		k_busy_wait(10000);
```

Comment:
Is the busy wait necessary?  Seems simpler to just let the thing loop.  It won't hurt to set the GPIO active repeatedly, will it?

Reply:
right